iree/base/atomic_slist.h

```diff
@@ -0,0 +1,264 @@
+// This is useful for when |list| is known to be inaccessible to any other
+// thread, such as when populating a stack-local list prior to sharing it.
+void iree_atomic_slist_push_unsafe(iree_atomic_slist_t* list,
+                                   iree_atomic_slist_entry_t* entry);
+
+// Pops the most recently pushed entry from the list and returns it.
+// Returns NULL if the list was empty at the time it was queried.
+//
+//   existing slist: C B A
+//  resulting slist: B A
+//   returned entry: C
+iree_atomic_slist_entry_t* iree_atomic_slist_pop(iree_atomic_slist_t* list);
+
+// Defines the approximate order in which a span of flushed entries is returned.
+typedef uint32_t iree_atomic_slist_flush_order_t;
+enum {
+  // |out_head| and |out_tail| will be set to a span of the entries roughly in
+  // the order they were pushed to the list in LIFO (stack) order.
+  //
+  // Example:
+  //    slist: C B A
+  //   result: C B A (or when contended possibly C A B)
+  IREE_ATOMIC_SLIST_FLUSH_ORDER_APPROXIMATE_LIFO,
+  // |out_head| and |out_tail| will be set to the first and last entries
+  // pushed respectively, turning this LIFO slist into a FIFO queue.
+  //
+  // Example:
+  //    slist: C B A
+  //   result: A B C (or when contended possibly B A C)
+  IREE_ATOMIC_SLIST_FLUSH_ORDER_APPROXIMATE_FIFO,
+};
+
+// Removes all items from the list and returns them in **APPROXIMATELY** the
+// |flush_order| requested. As there are no order guarantees there may be slight
+// transpositions of entries that were pushed from multiple processors or even
+// interleaved entries within spans of entries pushed with
+// iree_atomic_slist_concat.
+//
+// If |out_tail| is not required it can be omitted and this may avoid the
+// need for the flush to walk the list and touch each entry.
+//
+// Returns true if any items were present and false if the output list is empty.
+// Note that because atomic data structures can race it's possible for there to
+// both be something in the list prior to this call and something in the list
+// after the call and yet the return can still be false.
+bool iree_atomic_slist_flush(iree_atomic_slist_t* list,
+                             iree_atomic_slist_flush_order_t flush_order,
+                             iree_atomic_slist_entry_t** out_head,
+                             iree_atomic_slist_entry_t** out_tail);
+
+//==============================================================================
+// Typed wrapper generator for iree_atomic_slist_t
+//==============================================================================
+
+// Typed and named wrappers for making atomic slists easier to work with.
+//
+// Usage:
+//  typedef struct {
+//    int some_fields;
+//    iree_atomic_slist_intrusive_ptr_t slist_next;
+//    int more_fields;
+//  } my_type_t;
+//  IREE_TYPED_ATOMIC_SLIST_WRAPPER(my_type, my_type_t,
+//                                  offsetof(my_type_t, slist_next));
```

Comment:
Neat, these macros work basically like a C++ templated class but in C?

Reply:
yep! so you'll get nice type errors if you try to use the wrong types together and such.